### ISSUE TITLE
N64 Audio Timing

### ISSUE THREAD

**Issue by thecoreyburton** [TASVideos/BizHawk#711]
I'm not sure if this has been reported before (if it has, then you're welcome to close this issue) but music on the N64 core in Bizhawk experiences a desync in relation to the video. Usually this would go unnoticed as the music is often a background loop with no timing relevance.

The best example is to load up Super Smash Bros. (U) in Bizhawk and let it play the intro. After around 25 seconds, there is a section of the introduction where character names are listed and then short clips of them are show. When each name is displayed, there's an orchestral hit in the audio track. At first it aligns perfectly - but it gradually goes out of time (most noticably around Yoshi / Kirby). It gradually increases in this fashion as the intro goes on when compared to the console version. It's not game-exclusive, as the story to Banjo-Kazooie exhibits the same behavior, with the music beginning in time but by the end of the introduction things are happening several seconds earlier in the soundtrack when compared to the video.

It should be noted that the sound effects and any new sounds start in the correct place, it just seems that it's either a gradual desync in long-standing audio or it could be extra lag is generated by the core. I'm not sure the cause of this but after playing around with timing methods and settings to no avail, I thought it should be mentioned!

**Comment by vgturtle127** [comment]
@thecoreyburton This is actually an issue with the audio plugin, I believe. From what I remember, with the exception of one or two plugins, the Smash Bros. intro almost always goes out of sync because it isn't handled correctly. Based on the fact that you said "N64 core" instead of the plugin. I'm assuming that's what you mean, so feel free to correct me if I am wrong.

**Comment by thecoreyburton** [comment]
Good catch! Yes, your assumptions are correct, my use of the term "core" was vague and was just intended to reference the fact it was localized to the N64 emulation aspect of Bizhawk. Apologies for the miswording on my behalf.

**Comment by vgturtle127** [comment]
@thecoreyburton It's alright, just wanted to make sure I was assuming correctly. 😄

**Comment by vgturtle127** [comment]
This is a common issue with audio plugins, and I'm not sure how to best approach this without just being like "update all the plugins and use more accurate plugins".

Even Azimer's audio has this issue sometimes, so if anybody could shed some light on how to approach this that would be great.

**Comment by zeromus** [comment]
Please evaluate this situation on other emulators, with other audio plugins. If there's a specific plugin that works, or if we can update a plugin with fixes that make this work, then we'll change this bug to "update that plugin". That's the only way this gets resolved.

**Comment by thecoreyburton** [comment]
The current version of Project 64, with the default audio plugin, does not exhibit this problem. There is potentially more information over at #879 if need be.

**Comment by vgturtle127** [comment]
@thecoreyburton Just so you know, timing, in general, is an issue with most emulators and plugins, since most plugins share some code. An N64 emulator is just a CPU emulator with plugin functionality, essentially. Project64 is probably the best N64 emulator, especially in combination with certain plugins.

We use the standard Mupen64Plus SDL audio plugin, and unless we can find another plugin with support for the Mupen64Plus plugin API, we are pretty much out of luck. I don't think Azimer's audio plugin supports Mupen64Plus, but that would be the best one to port if any.
